### PR TITLE
Add `fullstack` tag for system specs

### PR DESCRIPTION
**Pull request by mjankowski**
This is sort of a followup on a bunch of previous PRs:

- https://github.com/mastodon/mastodon/pull/28682
- https://github.com/mastodon/mastodon/pull/28681
- https://github.com/mastodon/mastodon/pull/27932
- https://github.com/mastodon/mastodon/pull/27732

As noted in comments across those, I had a few spec coverage goals that are blocked by the current configuration of system specs:

- Continue to modernize some of our controller coverage, moving portions of it into request specs, and other portions into system specs (Likely driven by rack_test for a good chunk of the settings/admin ones); leaving behind only the things which are narrowly about the controller as a ruby class. I suspect this will take a while but will be good to go one chunk at a time, and will hopefully lead to some spec LOC reduction, but improvement in clarity and coverage.
- Add more full-stack coverage that exercises the react app -- this is a massive gap right now

Previous efforts have been rejected for various naming/terminology/organizational reasons. This one takes a slightly different approach to hopefully handle the issues from the prior ones. Summary of changes:

- Move all the system specs which were previously in `spec/features/` to `spec/system/`. This is very much a naming-only move to align with where current rspec/rails versions put things (misc reading: [1](https://railsnotes.xyz/blog/feature-specs-to-system-specs#), [2](https://rspec.info/blog/2017/10/rspec-3-7-has-been-released/#rails-actiondispatchsystemtest-integration-system-specs), [3](https://rspec.info/features/6-1/rspec-rails/system-specs/system-specs/))
- Move the system specs which were previously the only occupants of `spec/system/` into `spec/system/fullstack/` as a way of isolating them from the others (to enable next bullet...)
- Update rspec config to apply the `:fullstack` tag to all spec files in that directory.
- Update the capybara configuration to use `rack_test` as the default driver for system specs; and to use the configured `Capybara.javascript_driver` (currently headless chrome) for examples with that `:fullstack` tag.
- Update the configuration which was skipping all of spec/system by default (run in CI only) to still skip them, but to do it based on the rspec tag instead of an env var.
- Misc other config changes (CI command, rake task, streaming manager, js errors check, etc) all along those lines as well

Curious to get feedback on this direction, will leave a few inline comments re: future ideas.

### CODE REVIEW

File: .github/workflows/test-ruby.yml [hunk 1]
```diff
@@ -184,6 +184,8 @@ jobs:
       DISABLE_SIMPLECOV: true
       RAILS_ENV: test
       BUNDLE_WITH: test
+      LOCAL_DOMAIN: localhost:3000
```
Comment (mjankowski) [inline .github/workflows/test-ruby.yml:187]: One future change we should make would be to align these values for the regular spec suite and the fullstack system specs, so that we don't need to set them here. Ideally they'd be read from `.env.test` or something, and use same values across the whole suite.

File: spec/support/streaming_server_manager.rb [hunk 2]
```diff
@@ -125,6 +125,6 @@ def streaming_server_manager
   end
 
   def streaming_examples_present?
-    RUN_SYSTEM_SPECS
+    RSpec.world.filtered_examples.values.flatten.any? { |example| example.metadata[:fullstack] == true }
```
Comment (mjankowski) [inline spec/support/streaming_server_manager.rb:128]: Sort of verbose, but using same approach as the search manager class does. There may be a simpler way to check for the presence of an rspec tag...?

File: spec/rails_helper.rb [hunk 3]
```diff
@@ -26,12 +22,9 @@
 Dir[Rails.root.join('spec', 'support', '**', '*.rb')].each { |f| require f }
 
 ActiveRecord::Migration.maintain_test_schema!
-WebMock.disable_net_connect!(allow: Chewy.settings[:host], allow_localhost: RUN_SYSTEM_SPECS)
+WebMock.disable_net_connect!(allow: Chewy.settings[:host], allow_localhost: true)
```
Comment (ClearlyClaire) [inline spec/rails_helper.rb:25]: Can we somehow make it dependent on the tag? I think it makes sense to forbid any network connection on the non-fullstack tests. But I'm not super attached to it, I suppose this is fine too if the alternative is too much trouble.

**Review by ClearlyClaire — COMMENTED** [review]
Overall, I think this is a sensible approach. I'm curious if the extra setup for fullstack tests would cause issues for non-fullstack tests. i.e. can you do `bundle exec rspec --tag fullstack` and have it run all the tests without issues?

There currently are test failures on the CI, but I haven't investigated them.

File: spec/rails_helper.rb [hunk 4]
```diff
@@ -2,13 +2,9 @@
...
 
-# This needs to be defined before Rails is initialized
-RUN_SYSTEM_SPECS = ENV.fetch('RUN_SYSTEM_SPECS', false)
-
-if RUN_SYSTEM_SPECS
-  STREAMING_PORT = ENV.fetch('TEST_STREAMING_PORT', '4020')
-  ENV['STREAMING_API_BASE_URL'] = "http://localhost:#{STREAMING_PORT}"
-end
+# This needs to be defined before Rails is initialized (used by full stack system specs)
+STREAMING_PORT = ENV.fetch('TEST_STREAMING_PORT', '4020')
```
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply spec/rails_helper.rb:6]: That seems sensible to me.

File: .github/workflows/test-ruby.yml (hunk 1 above)
Comment (ClearlyClaire) [reply .github/workflows/test-ruby.yml:187]: I think that would make sense, but I'm also a bit wary of the extensive changes in the test suite this may require.

Comment (mjankowski) [reply .github/workflows/test-ruby.yml:187]: Yeah ... there are ~50 references to that `ngrok.io` host value across spec dirs. I suspect that some of that are using that just because it happens to be the value, not because the spec cares about the hostname. Maybe I'll do a quick spike branch on changing that to align the values and see how it goes.

File: spec/rails_helper.rb (hunk 3 above)
Comment (mjankowski) [reply spec/rails_helper.rb:25]: I was initially trying to change this constant to be set based on the rspec tags (the logic currently in streaming manager in this PR) instead of the env var, but the place it was initially set is before rspec loads so thats not available yet. I'll take a stab at moving it to be before this line but after rspec load, and see what that looks like.

**Comment by mjankowski** [comment]
> Overall, I think this is a sensible approach. I'm curious if the extra setup for fullstack tests would cause issues for non-fullstack tests. i.e. can you do `bundle exec rspec --tag fullstack` and have it run all the tests without issues?

Good question ... we'd have to look at both using the JS-browser for the previous system specs that use rack_test and see if they are ok under selenium (they were NOT on my quick initial checks in a previous PR); and we'd also have to look at whether having the streaming server running and connected in the background impacts those or any other specs as well. I haven't thoroughly looked at this, other than that initial check.

> There currently are test failures on the CI, but I haven't investigated them.

Yeah there's definitely some issue here. It's a little hard because we do have those intermittent failures across the selenium-driven system specs already ... but I'm also seeing these as intermittent local failures more so in this PR branch than regularly, and with different failures, so there's probably something going on from this branch. I'll work on that a bit.

Separately, for me locally, the oauth spec specifically always fails entirely. Not sure if there's a mac/linux issue where it fails for me and passes CI, or some other env/config thing I've got going where I don't realize the difference.

**Comment by mjankowski** [comment]
Think I worked out the flaky specs aspect ... I had forgotten to change the db cleaner config in a place it needed to be changed, and changed it in a place it did not need changing. Adjusted that, things passing now. Still looking at other stuff.

File: .github/workflows/test-ruby.yml (hunk 1 above)
Comment (mjankowski) [reply .github/workflows/test-ruby.yml:187]: Not terrible - https://github.com/mastodon/mastodon/compare/main...mjankowski:align-hostnames - this WIP branch has a few edge case failures where I'm not sure on best approach to resolve, but the diff here is not that bad. I'm going to call this resolved for purposes of this PR, but may nudge this forward separately.

**Review by ClearlyClaire — APPROVED** [review]

**Comment by ClearlyClaire** [comment]
I do appreciate the switch from an environment variable to a tag, but taking a step back, I find a sub-directory (`spec/system/fullstack`) having special behavior slightly weird, and I think the previous behavior of having this dependant on top-level directories made more sense.

Are you aware of any established convention around that in other Rails project?

**Comment by mjankowski** [comment]
> I do appreciate the switch from an environment variable to a tag, but taking a step back, I find a sub-directory (`spec/system/fullstack`) having special behavior slightly weird, and I think the previous behavior of having this dependant on top-level directories made more sense.

I also think it's weird, but added it here based on prior guidance.

> Are you aware of any established convention around that in other Rails project?

In 99% of every (recent-ish) Rails project I've looked at, the convention is to put all the system specs into `spec/system/` and use rspec tags to guide which configured capybara browser drives those examples, usually on a per-file basis. This is usually done with having `rack_test` as the default driver, and then configuring a `:js` or `:javascript` tag to use whatever the configured selenium JS-enabled browser is. I tried this in a previous PR and it was rejected.

The only special case we have on top of that is the spinning up/down of the streaming server via that manager class.

**Comment by ClearlyClaire** [comment]
Trying to sum up my thoughts about all this:
- I do think using a disabled-by-default tag rather than the weird mix from before is an improvement
- I think having such heavy specs in a separate directory does make sense, but having them in a sub-directory of `spec/system` feels a bit weird
- I guess we could have one tag for selenium and one tag for the streaming server (e.g. the admin system spec that test JS do not need the streaming server), but maybe that's overkill?

**Comment by mjankowski** [comment]
> * I do think using a disabled-by-default tag rather than the weird mix from before is an improvement

Great, agreed. 

> * I think having such heavy specs in a separate directory does make sense, but having them in a sub-directory of `spec/system` feels a bit weird
> * I guess we could have one tag for selenium and one tag for the streaming server (e.g. the admin system spec that test JS do not need the streaming server), but maybe that's overkill?

I think what you described is the most typical and intended way to do this ... use tags at the appropriate level (whole file, just a context, just one example, etc) to describe the features you want on/off/changed/whatever. We do this already for the paperclip processing and sidekiq behavior; so also doing it for the "I need a streaming server running" and "I need to use a JS browser" use cases feel normal to me. I tried this in a previous PR that was rejected.

With the appropriate tagging in place, you can then configure what should run by default (while preserving ability to run full) and what should only run on CI and so on.

Part of what I think might feel odd here is that since we *currently have* this hard separation by directory breaking that seems weird ... but in some hypothetical future when we migrate some of the current controller specs to become system specs, and ideally add many more new ones for some of the JS-requiring features it will not seem weird.

My own side note - I think it makes sense to use dirs for things like "admin", kind of like we have in the api request specs and in the controller specs now ... basically have it mirror either the actual primary code organization or some app-level concept; rather than for feature enabled/disable (which like you noted, is better done with tags).

**Comment by mjankowski** [comment]
Putting aside whatever env vars or other config, I think the CI command would just be like `bin/rspec --tag js --tag streaming` - which would run all the examples with either or both tags?

**Comment by ClearlyClaire** [comment]
> Putting aside whatever env vars or other config, I think the CI command would just be like `bin/rspec --tag js --tag streaming` - which would run all the examples with either or both tags?

But it would also run every other test, right?

**Comment by mjankowski** [comment]
> > Putting aside whatever env vars or other config, I think the CI command would just be like `bin/rspec --tag js --tag streaming` - which would run all the examples with either or both tags?
> 
> But it would also run every other test, right?

If I understand the question, then yes it would still run everything on CI. What I was envisioning on CI was to preserve the current approach:

- Run plain `bin/rspec` with no args/tags/options for the main `test` step. Would run everything except what we configure to exclulde by default (`js` and `streaming` in our hypothetical, in addition to already-excluded `search`)
- Run `bin/rspec --tag js --tag streaming` for the `test-e2e` step, which would run anything tagged with one/both of those, and which would need to spin up the streaming server and use headless chrome.
- Leace the search one as-is -- this already uses a tag approach for exclusion and the search data manager.

For local dev, you'd have the same default behavior of excluding the defaults, but could opt in with the tags if desired.

If that was not what you meant ... then I don't know what you meant.

**Comment by ClearlyClaire** [comment]
Sorry, I misunderstood how `--tag` worked, `bin/rspec --tag js --tag streaming` would do exactly what we want (running only tests that involve either tag).

**Comment by mjankowski** [comment]
Should I reformulate this into one that includes both the file dir consolidation and the tag approach into another PR and we can review that?

**Comment by ClearlyClaire** [comment]
What do you mean by the file dir consolidation?

**Comment by mjankowski** [comment]
> What do you mean by the file dir consolidation?

Moving the system specs which are currently in `spec/features/` into `spec/system/` (as done in this PR).

I could probably start a branch from this PR's branch, and keep the file moves, but replace the `fullstack` dir/tag approach with a pure tag approach using js/streaming, and see how that looks?

**Comment by ClearlyClaire** [comment]
Yeah, sounds like a plan to me!

**Comment by mjankowski** [comment]
Closing in favor of https://github.com/mastodon/mastodon/pull/30206
